Stores/src/main/java/AppUpdateWorker.kt

```diff
@@ -0,0 +1,94 @@
+import androidx.work.*
+import com.google.common.util.concurrent.ListenableFuture
+import com.infomaniak.lib.core.utils.SentryLog
+import com.infomaniak.lib.stores.updatemanagers.WorkerUpdateManager
+import io.sentry.Sentry
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+import java.util.concurrent.TimeUnit
+
+class AppUpdateWorker(appContext: Context, params: WorkerParameters) : ListenableWorker(appContext, params) {
+
+    private val updateManager by lazy { WorkerUpdateManager(appContext) }
+
+    override fun startWork(): ListenableFuture<Result> {
+        SentryLog.i(TAG, "Work started")
+
+        return CallbackToFutureAdapter.getFuture { completer ->
+            updateManager.installDownloadedUpdate(
+                onInstallSuccess = { completer.setResult(Result.success()) },
+                onInstallFailure = { exception ->
+                    // This avoid the user being instantly reprompted to download update
+                    Sentry.captureException(exception)
+                    completer.setResult(Result.failure())
+                },
+            )
+        }
+    }
+
+    private fun CallbackToFutureAdapter.Completer<Result>.setResult(result: Result) {
+        set(result)
+        SentryLog.d(TAG, "Work finished")
+    }
+
+    class Scheduler(appContext: Context) {
+
+        private val workManager = WorkManager.getInstance(appContext)
+        private val storesLocalSettings = StoresLocalSettings.getInstance(appContext)
+
+        fun scheduleWorkIfNeeded() {
+            if (BuildConfig.HAS_GPLAY_SERVICES && storesLocalSettings.hasAppUpdateDownloaded) {
+                SentryLog.d(TAG, "Work scheduled")
+
+                val workRequest = OneTimeWorkRequestBuilder<AppUpdateWorker>()
+                    .setConstraints(Constraints.Builder().setRequiresBatteryNotLow(true).build())
+                    // We start with a delayed duration, so that when the app quickly come back to foreground because the user
+                    // was just switching apps, the service is not launched
+                    .setInitialDelay(INITIAL_DELAY, TimeUnit.SECONDS)
+                    .build()
+
+                workManager.enqueueUniqueWork(TAG, ExistingWorkPolicy.KEEP, workRequest)
+            }
+        }
+
+        suspend fun cancelWorkIfNeeded() = withContext(Dispatchers.IO) {
+
+            val workInfo = workManager.getWorkInfos(
+                WorkQuery.Builder.fromUniqueWorkNames(listOf(TAG))
+                    .addStates(listOf(WorkInfo.State.BLOCKED, WorkInfo.State.ENQUEUED))
+                    .build(),
+            ).get()
+
+            workInfo.forEach {
+                workManager.cancelWorkById(it.id)
+                SentryLog.d(TAG, "Work cancelled")
+            }
+        }
+    }
+
+    companion object {
+        private const val TAG = "AppUpdateWorker"
+        private const val INITIAL_DELAY = 10L // 10s
```

Comment:
What I like to do is always indicate the unit in the variable name, like so: `INITIAL_DELAY_SECONDS`. That way, you don't have to look at the constant declaration to know what's the unit. Not mandatory. If you prefer that way, it's also fine.

Reply:
You're right it's better